### PR TITLE
fix: Do not register clicks outside displayed config button

### PR DESCRIPTION
**Pull request by Mcrtin**
this is a complete rework of the ConfigButton -now called ConfigTile. This aims to fix the missing bounding box of the ConfigOptionElements.

### CODE REVIEW

**Comment by magicus** [comment]
What is the rationale for this change?

**Comment by Mcrtin** [comment]
the config button (for example the enabled button) spans the whole config tile, only leaving room for the reset button. it should instead only register clicks when the actual button is clicked.

**Comment by magicus** [comment]
Ok, I see. I suggest you rename this PR along those lines: `fix: Do not register clicks outside config button text`.

I'll review this after I'm finished with beta stuff.

**Comment by magicus** [comment]
I started reviewing this. From the look of it, this will change how the config screen is rendered, not just how clicks are registered? The old code had a lot of calls to e.g. `RenderUtils.drawRoundedRectWithBorder`. I can't see that this is present in the new code. 

Does this mean the look of the config screen has changed? If so, can you post a screenshot? If not, can you explain how you keep the same look -- what am I missing?

**Comment by Mcrtin** [comment]
there was a duplicate code. The `GeneralSettingsButton` does all the rendering; the same way the element classes did, but the settings screen looks a bit different because I changed the width and height a bit.
Before:
image
image
After:
image
image

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        poseStack.pushPose();
+        poseStack.translate(0f, (height - FontRenderer.getInstance().getFont().lineHeight + 8) / 2f - 5, 0f);
+        super.renderWidget(poseStack, mouseX, mouseY, partialTick);
+        poseStack.popPose();
+    }
+
+    protected void onUpdate(String text) {
+        Object parsedValue = configHolder.tryParseStringValue(text);
+        if (parsedValue != null) {
+            if (!Objects.equals(parsedValue, configHolder.getValue())) configHolder.setValue(parsedValue);
```
Comment (magicus) [inline common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java]: Please always use braces in if statements, except for trivial return/break/etc.

File: common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    protected CustomColor getBackgroundColor() {
+        return configHolder.valueChanged() ? super.getBackgroundColor() : BACKGROUND_COLOR;
+    }
+
+    @Override
+    public void playDownSound(SoundManager handler) {
+        if (configHolder.valueChanged()) super.playDownSound(handler);
```
Comment (magicus) [inline common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java]: Same here; use braces.

**Review by magicus — CHANGES_REQUESTED** [review]
I think this looks okay, apart from the "if" changes. But I will need bolyai to look at it as well, this is his code originally.

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java (hunk 1 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java]: he said it'd be ok if they're on one line, referring to this blog post: https://www.codecentric.de/wissens-hub/blog/curly-braces

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java]: No, this is not okay. I might not have been clear, "if it fits the line", but this borderline doesn't. So you just add the braces.

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    public void renderWidget(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        poseStack.pushPose();
+        poseStack.translate(0f, (height - FontRenderer.getInstance().getFont().lineHeight + 8) / 2f - 5, 0f);
+        super.renderWidget(poseStack, mouseX, mouseY, partialTick);
+        poseStack.popPose();
+    }
+
+    protected void onUpdate(String text) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: You don't have to invent this whole abstract/overriden method if you added this method as a consumer in the constuctor..

File: common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java (hunk 2 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java]: Use it.

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java (hunk 3 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: how can i do that? i tried `this::onUpdate` but I "Cannot reference 'this' before supertype constructor has been called"

File: common/src/main/java/com/wynntils/screens/settings/widgets/EnumSettingsButton.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public void onPress() {
+        // we use instead AbstractWidget#mouseClicked because we also want to have an action on the right mouse button
+    }
+
+    private static <E extends Enum<E>> int getWidth(Type type) {
+        Font font = FontRenderer.getInstance().getFont();
+        return EnumSet.allOf((Class<E>) type).stream()
```
Comment (magicus) [inline common/src/main/java/com/wynntils/screens/settings/widgets/EnumSettingsButton.java]: Is this how Spotless format this statement? 😱 

I think you need to break it down. Can you extract the stream operation into a local variable `maxWidth` or such, and then return it + 8.

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java (hunk 3 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: What if you were to just pass it as a lambda, would that have worked? I'm not sure it's better, but maybe?

Or you could make this a private static helper function, and pass it the configHolder as well as argument, and call this as a lambda from the constructor. I'm pretty sure that is allowed.

(I agree with bolyai that it is better if we don't need this override when there is already a mechanism for setting a callback)

Comment (Mcrtin) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: I tried everything before i resorted to making an abstract method but to invoke `setRenderColor` I have to have the reference to the object.

Comment (magicus) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: I see. Ok, your explanation sounds reasonable to me. I'll leave it to bolyai to resolve this conversation if he's happy with the reply.

**Review by kristofbolyai — COMMENTED** [review]
You might have realized this, but to be honest, I don't like this PR. We basically have an almost 500 line PR which touches an essential system, the config system GUI, a completely "rewritten" widget just to fix a small "bug". 

This is kind of a "if it is don't broken, don't fix it" approach from me, and I am also slightly paranoid it will accidentally break something.

Nevertheless, I will take a close look at this tomorrow, and try to integrate this (after the necessary reviews).

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java (hunk 3 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: Okay, I see. Let's keep it.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java:27]: You can resolve this after you add the `@Override` annotation for it.

File: common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+package com.wynntils.screens.settings.widgets;
+
+import com.wynntils.core.config.ConfigHolder;
+import com.wynntils.screens.base.TextboxScreen;
+import com.wynntils.screens.base.widgets.TextInputBoxWidget;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.render.FontRenderer;
+
+public class TextInputBoxSettingsWidget extends TextInputBoxWidget {
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/TextInputBoxSettingsWidget.java]:
```suggestion
```

File: common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+package com.wynntils.screens.settings.widgets;
+
+import com.wynntils.core.config.ConfigHolder;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.colors.CustomColor;
+import java.util.List;
+import net.minecraft.client.sounds.SoundManager;
+import net.minecraft.network.chat.Component;
+
+final class ResetButton extends GeneralSettingsButton {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java]:
```suggestion
private final class ResetButton extends GeneralSettingsButton {
```

File: common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                12,
+                Component.translatable("screens.wynntils.settingsScreen.reset.name"),
+                List.of(Component.translatable("screens.wynntils.settingsScreen.reset.description")));
+        this.configHolder = configHolder;
+        this.onClick = onClick;
+    }
+
+    @Override
+    protected CustomColor getTextColor() {
+        return !configHolder.valueChanged() ? CommonColors.GRAY : CommonColors.WHITE;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/ResetButton.java]: Can you flip this (to remove `!`)?

File: common/src/main/java/com/wynntils/screens/settings/widgets/EnumSettingsButton.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import com.wynntils.utils.render.FontRenderer;
+import java.lang.reflect.Type;
+import java.util.EnumSet;
+import java.util.List;
+import net.minecraft.client.gui.Font;
+import net.minecraft.network.chat.Component;
+import org.lwjgl.glfw.GLFW;
+
+public class EnumSettingsButton<E extends Enum<E>> extends GeneralSettingsButton {
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/EnumSettingsButton.java]:
```suggestion
```

File: common/src/main/java/com/wynntils/screens/settings/widgets/CloseButton.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+package com.wynntils.screens.settings.widgets;
+
+import com.wynntils.screens.settings.WynntilsBookSettingsScreen;
+import com.wynntils.utils.render.Texture;
+import java.util.List;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+
+public class CloseButton extends GeneralSettingsButton {
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/CloseButton.java]:
```suggestion
```

File: common/src/main/java/com/wynntils/screens/settings/widgets/BooleanSettingsButton.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import com.wynntils.core.config.ConfigHolder;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.colors.CustomColor;
+import com.wynntils.utils.render.FontRenderer;
+import java.util.List;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+
+public class BooleanSettingsButton extends GeneralSettingsButton {
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/BooleanSettingsButton.java]:
```suggestion
```

File: common/src/main/java/com/wynntils/screens/settings/widgets/ApplyButton.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.wynntils.core.components.Managers;
+import com.wynntils.screens.settings.WynntilsBookSettingsScreen;
+import com.wynntils.utils.render.Texture;
+import java.util.List;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+
+public class ApplyButton extends GeneralSettingsButton {
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/settings/widgets/ApplyButton.java]:
```suggestion
```

File: common/src/main/java/com/wynntils/screens/base/widgets/TextInputBoxWidget.java [hunk 12]
```diff
@@ -583,4 +583,9 @@ private void replaceHighlighted(String text) {
         this.setHighlightPosition(this.cursorPosition);
         this.onUpdateConsumer.accept(this.textBoxInput);
     }
+
+    /**
+     * if there is no on update consumer given in the constructor this method gets called instead.
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/base/widgets/TextInputBoxWidget.java]:
```suggestion
     * If there is no on update consumer given in the constructor, this method gets called instead.
```

**Comment by Mcrtin** [comment]
The Tooltip of the categories is rendered under the config buttons:
image
how can I fix that? Is that an issue with the tooltips? It works with the tooltips of the config buttons:
image

**Comment by kristofbolyai** [comment]
It is an unrelated issue to this PR, you didn't cause it. I will check if it requires a lot of stuff to be rewritten..

**Review by magicus — APPROVED** [review]

**Review by kristofbolyai — APPROVED** [review]
Okay, this looks sane. Thank you for holding on, this took a while.

**Pull request merged**